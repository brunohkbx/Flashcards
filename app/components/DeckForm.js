import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { HelperText, TextInput } from 'react-native-paper';
import { View } from 'react-native';

export const FormikDeckForm = ({ handleSubmit, initialValues, formRef }) => {
  const validationSchema = Yup.object().shape({
    title: Yup.string().required('Required')
  });

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      enableReinitialize
      validateOnBlur={false}
      ref={formRef}
      component={DeckForm}
      onSubmit={(values, actions) => handleSubmit(values, actions)}
    />
  );
};

export const DeckForm = props => {
  const {
    handleChange,
    handleBlur,
    touched,
    errors,
    values,
  } = props;

  return (
    <View>
      <TextInput
        label="Title"
        value={values.title}
        onChangeText={handleChange('title')}
        onBlur={handleBlur('title')}
        error={!!(touched.title && errors.title)}
        autoFocus
      />
      <HelperText type="error" visible={!!(touched.title && errors.title)}>
        {errors.title}
      </HelperText>
    </View>
  );
};

FormikDeckForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object,
  formRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
};

FormikDeckForm.defaultProps = {
  initialValues: { title: '', questions: [] }
};

export default FormikDeckForm;
