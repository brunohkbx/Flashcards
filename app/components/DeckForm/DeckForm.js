import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Formik, FieldArray } from 'formik';
import * as Yup from 'yup';
import uuid from 'uuid';
import { View, KeyboardAvoidingView } from 'react-native';
import FAB from '../FAB';
import BottomRightContainer from '../BottomRightContainer';
import { Header } from 'react-navigation';
import FormContent from './FormContent';

export const FormikDeckForm = ({ handleSubmit, initialValues, formRef }) => {
  const validationSchema = Yup.object().shape({
    title: Yup.string().required('Required'),
    questions: Yup.array().of(
      Yup.object().shape({
        question: Yup.string().required('Required'),
        answer: Yup.string().required('Required'),
        id: Yup.string().required('Required')
      })
    )
  });

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      enableReinitialize
      validateOnBlur={false}
      ref={formRef}
      render={formikProps => (
        <FieldArray name="questions" component={DeckForm} />
      )}
      onSubmit={(values, actions) => handleSubmit(values, actions)}
    />
  );
};

export class DeckForm extends Component {
  addNewFlashcard = () => {
    const { push } = this.props;

    push({ question: '', answer: '', id: uuid() });
  }

  handleFlashcardDeleted = index => {
    const { remove } = this.props;

    remove(index);
  }

  render() {
    return (
      <View style={{flex: 1, padding: 8}}>
        <KeyboardAvoidingView
          style={{flex: 1}}
          behavior="padding"
          enabled
          keyboardVerticalOffset={Header.HEIGHT + 35}
        >
          <FormContent
            onFlashcardDeleted={this.handleFlashcardDeleted}
            {...this.props.form}
          />
        </KeyboardAvoidingView>
        <BottomRightContainer right={16} bottom={16} >
          <FAB handlePress={this.addNewFlashcard} />
        </BottomRightContainer>
      </View>
    );
  }
}

FormikDeckForm.propTypes = {
  initialValues: PropTypes.shape({
    title: PropTypes.string.isRequired,
    questions: PropTypes.array.isRequired,
    id: PropTypes.string.isRequired
  }),
  formRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  handleSubmit: PropTypes.func.isRequired,
};

export default FormikDeckForm;
