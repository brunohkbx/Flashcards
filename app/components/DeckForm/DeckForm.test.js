import React from 'react';
import { shallow } from 'enzyme';
import { FormikDeckForm, DeckForm } from './DeckForm';

describe('FormikDeckForm', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      handleSubmit: jest.fn(),
    }, propOverrides);

    const wrapper = shallow(<FormikDeckForm {...defaultProps} />);

    const submit = () => wrapper.simulate('submit');

    return {
      wrapper,
      submit
    };
  };

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  test('onSubmit calls handleSubmit', () => {
    const mockHandleSubmit = jest.fn();
    const { submit } = setup({ handleSubmit: mockHandleSubmit });

    submit();

    expect(mockHandleSubmit).toHaveBeenCalled();
  });
});

describe('DeckForm', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      handleChange: jest.fn(() => jest.fn),
      handleBlur: jest.fn(() => jest.fn),
      touched: {},
      errors: {},
      values: {},
    }, propOverrides);

    const wrapper = shallow(<DeckForm {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    const pressOnFab = () => {
      wrapper
        .find('FAB')
        .dive()
        .simulate('press');
    };

    return {
      wrapper,
      wrapperInstance,
      pressOnFab
    };
  };

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  test('FAB onPress calls addNewQuestion', () => {
    const { pressOnFab, wrapperInstance } = setup();
    const spy = jest
      .spyOn(wrapperInstance, 'addNewQuestion')
      .mockImplementation(() => jest.fn());

    pressOnFab();

    expect(spy).toHaveBeenCalled();
  });

  describe('addNewQuestion', () => {
    it('pushes a new empty question into formik values', () => {
      const values = { questions: [] };
      const mockSetFieldValue = jest.fn();
      const { wrapperInstance } = setup(
        { values,
          setFieldValue: mockSetFieldValue
        }
      );

      wrapperInstance.addNewQuestion();

      expect(
        mockSetFieldValue
      ).toHaveBeenCalledWith('questions.0.question', '');

      expect(
        mockSetFieldValue
      ).toHaveBeenCalledWith('questions.0.answer', '');

      expect(
        mockSetFieldValue
      ).toHaveBeenCalledWith('questions.0.id', expect.any(String));
    });
  });

  describe('handleRemoveFlashcard', () => {
    it('removes the given question from formik values', () => {
      const values = { questions: [
          { id: '1' },
          { id: '2' }
        ]
      };

      const mockSetFieldValue = jest.fn();

      const { wrapperInstance } = setup(
        { values,
          setFieldValue: mockSetFieldValue
        }
      );

      wrapperInstance.handleRemoveFlashcard('1');

      expect(
        mockSetFieldValue
      ).toHaveBeenCalledWith('questions', [{ id: '2' }]);
    });
  });

  describe('setScrollViewRef', () => {
    it('assigns a ref to this.scrollView', () => {
      const mockScrollView = jest.fn();
      const { wrapperInstance } = setup();

      wrapperInstance.setScrollViewRef(mockScrollView);

      expect(wrapperInstance.scrollView).toEqual(mockScrollView);
    });
  });

  describe('handleContentSizeChange', () => {
    it('calls scrollToEnd on scrollView ref', () => {
      const mockScrollToEnd = jest.fn();
      const { wrapperInstance } = setup();
      wrapperInstance.scrollView = { scrollToEnd: mockScrollToEnd };

      wrapperInstance.handleContentSizeChange();

      expect(mockScrollToEnd).toHaveBeenCalledWith({ animated: true });
    });
  });
});
