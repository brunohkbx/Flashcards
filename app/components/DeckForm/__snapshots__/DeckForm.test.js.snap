// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckForm renders properly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "padding": 8,
    }
  }
>
  <KeyboardAvoidingView
    behavior="padding"
    enabled={true}
    keyboardVerticalOffset={99}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <FormContent
      handleChange={[MockFunction]}
      onFlashcardDeleted={[Function]}
      values={Object {}}
    />
  </KeyboardAvoidingView>
  <Styled(Component)
    bottom={16}
    right={16}
  >
    <withTheme(FAB)
      handlePress={[Function]}
      icon="credit-card"
    />
  </Styled(Component)>
</Component>
`;

exports[`FormikDeckForm renders properly 1`] = `
<Formik
  enableReinitialize={true}
  isInitialValid={false}
  onSubmit={[Function]}
  render={[Function]}
  validateOnBlur={false}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": false,
      "_nodes": Array [
        "questions",
        "title",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "questions": ArraySchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_subType": ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": false,
            "_nodes": Array [
              "id",
              "answer",
              "question",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "answer": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": false,
                },
                "_mutate": false,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "id": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": false,
                },
                "_mutate": false,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "question": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": false,
                },
                "_mutate": false,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          },
          "_type": "array",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
        "title": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
/>
`;
